### PR TITLE
Add Meson build system support

### PR DESCRIPTION
**Pull request by LonerDan**
Adds support for Meson build system as an alternative to the Makefile.
This adds more flexibility for build setups that may not be the same as the
makefile expects and also allows for example simpler integration into Buildroot.
Changes to the source files were kept to the minimum, only parametrizing
some hardcoded paths.
If this commit is accepted, the project could also be added to the Meson WrapDB
as a next step.

Signed-off-by: Daniel Brát <danek.brat@gmail.com>

### CODE REVIEW

**Comment by tlsa** [comment]
Hi @LonerDan,

> Adds support for Meson build system as an alternative to the Makefile.

I'm really sorry but I do not want to maintain a second buildsystem, so I will not merge the meson stuff.

> This adds more flexibility for build setups that may not be the same as the
> makefile expects and also allows for example simpler integration into Buildroot.

I am interested in improving the existing buildsystem to support easier integration into meta-buildsystems like Buildroot, Buildstream, Yocto, etc.  What specifically is inflexible at the moment?

File: test/units/file.c [hunk 1]
```diff
@@ -13,6 +13,12 @@
 #include "ttest.h"
 #include "test.h"
 
+#ifndef SAVE_FILE_DEST
+#define SAVE_FILE_DEST "build"
+#endif
+
+#define SAVE_FILE_PATH(filename) (SAVE_FILE_DEST "/" filename)
```
Comment (tlsa) [inline test/units/file.c:20]: This change to the save test paths looks good.  I'd merge this bit separately.

**Comment by danielealbano** [comment]
@tlsa, I am going to throw it out there, but why not actually "switch" to a different build platform like cmake that would simplify the integration of libcyaml?
I know that automake and in general makefiles are widely used in the unix world but cmake is almost "the standard" nowadays, it really brings on a lot of advantages and open a lot of doors when it comes to integrations.

If you would be open to consider it, I would be more than happy to rewrite it out.

@LonerDan if you need to use it with the Meson build system you may want to wrap it in cmake, I wrote a very simple wrapper for my project https://github.com/danielealbano/cachegrand/blob/main/3rdparty/libcyaml.cmake

**Comment by tlsa** [comment]
@danielealbano

> I am going to throw it out there, but why not actually "switch" to a different build platform like cmake that would simplify the integration of libcyaml?

I'm still interested to know how the current buildsystem is difficult to integrate?

I've seen it ported to [Arch](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=libcyaml), [Debian](https://sources.debian.org/src/libcyaml/1.2.0-2/debian/rules/), [FreeBSD](https://svnweb.freebsd.org/ports/head/textproc/libcyaml/Makefile?view=markup), [Nix](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/libcyaml/default.nix#L22), [Yocto](https://github.com/schnitzeltony/meta-musicians/blob/master/recipes-support/libcyaml/libcyaml.bb), [and more](https://repology.org/project/libcyaml/versions), and none of their build scripts look particularly complicated or different from other packages.

> I know that automake and in general makefiles are widely used in the unix world but cmake is almost "the standard" nowadays

That varies from community to community.  Gnome has shifted towards Meson.  Everyone will have some buildsystem they're most familiar with.

> it really brings on a lot of advantages and open a lot of doors when it comes to integrations.

Please give me more details!

> If you would be open to consider it, I would be more than happy to rewrite it out.

There would have to be huge advantages for me. At the moment, I like what I have, because:

1. I understand it.
2. I'm familiar with how to use it.
3. It does what I need.
4. Simple: "Just run `make`."
5. Supports loads of testing variants:
   * `make test`
   * `make test-quiet`
   * `make test-verbose`
   * `make test-debug`
6. Any of the above under `valgrind`:
   * `make valgrind`
   * `make valgrind-quiet`
   * `make valgrind-verbose`
   * `make valgrind-debug`
7. `VARIANT=release` and `VARIANT=debug` builds for lib, and all test/valgrind builds.
8. Sanitizer builds and tests with `VARIANT=san`.
9. Run any of the above testing/valgrind forms with `TESTLIST=foo` to provide an allow list of tests to run.
10. Coverage reports in various output formats: `make coverage`.
11. A documentation target: `make docs`.

Off the top of my head.  And I use all that while developing. I also like that make is a pretty lightweight build dependency.

**Comment by danielealbano** [comment]
@tlsa I think your expectation is to build a library for which there are going to be packages that will be available in the OS repositories where mine is to be able to use a library with a number of useful functionalities implemented in the latest releases because the library moves quickly.

With cmake would be easier to get libcyaml built as static library and integrated whereas the current build system, as being "custom" **(1)**, makes it impossible and requires wrappers. With cmake you would also be able to build packages for the various distro as needed, actually simplifying the process as cmake already provides a number of functionalities that handle the process almost/entirely end to end and work for any OS and any platform out there without any additional effort.

Let me make a practical example.
Debian provides 1.2.0, ubuntu 1.1.0, the latest one is 1.2.1 and I personally need that, if you introduce the binary support (mentioning it just as example from another pr) and release a 1.2.2 the only way to use that functionality would be to have libcyaml as static library built out from the latest build.
With the current process the expectation is that the package maintainer will do a new build and then submits it but, sadly (and I love this lib!) it's not as widely used as libyaml itself and therefore the package updates are behind.
If the build system would in cmake switching to a newer version would be as is as: adding libcyaml as submodule + use the find_package function in cmake, that's it. It would also make possible to use libcyaml from Meson ( https://mesonbuild.com/CMake-module.html )

Talking about your current build system, cmake has the concept of "targets" whereas a target can be a library or an executable to build, a custom command, a test, etc.
In your specific case you would have a cmake file with one target, your library, plus one target for the tests, one custom command for the coverage, although the coverage is just running gcov/cov/other-cov to collect the code coverage stats, and one custom command for valgrind, both with a dependency on building the library and the tests.

The commands you provided above are custom-ish and they work well when you build the process manually, but cmake as mentioned provides a number of facilities to build packages for the distros directly, making it easier for the package maintainers.

Here a few examples:
- deb: https://cmake.org/cmake/help/latest/cpack_gen/deb.html
- rpm: https://cmake.org/cmake/help/latest/cpack_gen/rpm.html
- FreeBSD, DragonflyBSD, NetBSD, OpenBSD: https://cmake.org/cmake/help/latest/cpack_gen/freebsd.html
- productbuild (mac os x): https://cmake.org/cmake/help/latest/cpack_gen/productbuild.html
- bundle (mac os x): https://cmake.org/cmake/help/latest/cpack_gen/bundle.html
- nuget (windows): https://cmake.org/cmake/help/latest/cpack_gen/nuget.html
- archive (for ad hoc builds without a package manager): https://cmake.org/cmake/help/latest/cpack_gen/archive.html

All of them use Cpack that is a module to build packages for cmake, there are quite a few more.

And, among other things, by default it let you to do out-of-tree builds that are extremely useful.

For example
`VARIANT=release make`

would become the first time (and every other time in which you add a source file)

```
mkdir cmake-build-release
cd cmake-build-release
cmake .. -DCMAKE_BUILD_TYPE=Release
make
```

And for all the subsequent builds simply `make`

You can rewrite as one-liner
`mkdir cmake-build-release && cmake -S . -B cmake-build-release -DCMAKE_BUILD_TYPE=Release && cmake --build cmake-build-release`

I see great value when "common" tools / build systems are used, it makes easier to integrate them and makes possible to focus on actually solving problems instead of just working on trying to glue things together.

This is of course my personal opinion that might be greatly different from yours which I do respect very much :)

**(1)** With "custom" I mean e.g. you can't interface externally with it, e.g. you can't enumerate the build options and select the ones you need unless you know exactly the name of these and therefore you need some "code" to retrieve and/or pass them, etc.

**Comment by tlsa** [comment]
> @tlsa I think your expectation is to build a library for which there are going to be packages that will be available in the OS repositories where mine is to be able to use a library with a number of useful functionalities implemented in the latest releases because the library moves quickly.

Yes, this is fair.

> If the build system would in cmake switching to a newer version would be as is as: adding libcyaml as submodule + use the find_package function in cmake, that's it. It would also make possible to use libcyaml from Meson ( https://mesonbuild.com/CMake-module.html )

I used libcyaml in an application called [Bloodview](https://github.com/CodethinkLabs/bloodlight-firmware/tree/master/host/bloodview) and in that project I added a git submodule for libcyaml, and [two lines](https://github.com/CodethinkLabs/bloodlight-firmware/blob/master/host/Makefile#L105-L106) to the Makefile to depend on and build the static library.

> Talking about your current build system, cmake has the concept of "targets" whereas a target can be a library or an executable to build, a custom command, a test, etc.
> In your specific case you would have a cmake file with one target, your library, plus one target for the tests, one custom command for the coverage, although the coverage is just running gcov/cov/other-cov to collect the code coverage stats, and one custom command for valgrind, both with a dependency on building the library and the tests.

> The commands you provided above are custom-ish and they work well when you build the process manually, but cmake as mentioned provides a number of facilities to build packages for the distros directly, making it easier for the package maintainers.

> For example
> `VARIANT=release make`

> would become the first time (and every other time in which you add a source file)

> ```
> mkdir cmake-build-release
> cd cmake-build-release
> cmake .. -DCMAKE_BUILD_TYPE=Release
> make
> ```

At the moment `make VARIANT=release` or `make VARIANT=debug` override the default, and the default is to build the release if the release version flag is set.  So people building a tag or from a release tarball get the release version just from running `make`, and if you're on some random git revision, you get a debug build by default.

> I see great value when "common" tools / build systems are used, it makes easier to integrate them and makes possible to focus on actually solving problems instead of just working on trying to glue things together.

I understand your points and thank you for offering to convert to CMake, however, I'd rather stick to what I know and like using or it will just make me reluctant to work on this project!

**Comment by tlsa** [comment]
@LonerDan

> If this commit is accepted, the project could also be added to the Meson WrapDB as a next step.

I've just looked at https://mesonbuild.com/Using-the-WrapDB.html and it says:

> The Wrap database is a web service that provides Meson build definitions to projects that do not have it natively.

So from what I can tell, it's a database of Meson build instructions for projects which don't use Meson, so I think you should be able to submit it there, even though this project will stick to a Makefile upstream.

@danielealbano  I'm not sure if a similar database exists for CMake, but if so, you could probably put your CMake wrapper there?

**Comment by tlsa** [comment]
@LonerDan The .meson files for the tests look very complicated, I'm not sure why there's so much logic in there, but I think I'd need a test for the test logic to check all the tests I expect are actually run.

In any case, I am going to close this, and I suggest adding libcyaml to the Meson WrapDB instead. Thanks for your interest in libcyaml.

**Comment by eli-schwartz** [comment]
> @LonerDan The .meson files for the tests look very complicated, I'm not sure why there's so much logic in there, but I think I'd need a test for the test logic to check all the tests I expect are actually run.

It looks like an extremely complicated and unnecessary way of doing individual subtest reporting. Normally, one would just... run the test binary once, rather than running the test binary many times, each time with the name of a `static bool test_XXXXX` function.

Subtest reporting is probably better done via http://testanything.org/ (which meson does support, as do a number of other test harnesses or report collectors).

With my meson developer hat on, I don't think it's worth the meson.build complexity shown here in this PR.

@LonerDan BTW what happened to trying to submit it to the WrapDB? :) Someone is in the process of adding libyaml itself at https://github.com/mesonbuild/wrapdb/pull/254 and also wants libcyaml availability...

**Comment by eli-schwartz** [comment]
> There would have to be huge advantages for me. At the moment, I like what I have, because:

@tlsa 

I cannot judge for you whether any build system satisfies the most important requirement of all, "I the software author understand how to use it".

Regarding your additional points:
- meson has out of the box support for the following during configuration as options to `meson setup`:
  - release/debug: `--buildtype=release`
  - sanitizer via `-Db_sanitize=address,undefined`
  - static/dynamic via `-Ddefault_library=both`
  - coverage reports via `-Db_coverage=true` enables `ninja coverage` equivalent to `make coverage`
- test options
  - valgrind support via `meson test --wrap=valgrind`
  - running only certain tests:
    - if you describe the names of available tests in meson.build, `meson test testname1 testname2`
    - by passing arbitrary arguments to the single test binary, `meson test --test-args "testname1 testname2"`

valgrind, asan, running specific tests, can all be used in combination with each other (remembering that asan specifically must be set at configure time to ensure the tests are built with the correct CFLAGS/LDFLAGS).

Docs are simple enough (in meson, cmake, autotools, or plain old Make), just define a custom build target that runs the doxygen command.

meson does require python, so it is hardly lightweight, and even though there is [a lightweight port to C99](https://sr.ht/~lattis/muon/), it still requires two programs: one to configure, one to make
So, that is a lot less simple than "just run `make`.

I would caution against cmake, because:
- I am a developer of a cmake competitor ;)
- but also because cmake will inevitably lead to writing a fully turing-complete build description in "the cmake language", which is [a valid ray tracing language](https://64.github.io/cmake-raytracer/), and the cmake language probably cannot be independently reimplemented in a robust fashion. It is certainly the last thing to call "simple".
- cmake claims that pkg-config is bad/unreliable, and recommends instead using "cmake config" files that only cmake understands how to use, because they are written in that turing-complete ray tracing language referred to in the previous point.

You mentioned simplicity???

**Pull request closed without merging**